File: src/main/java/io/playground/orderservice/domain/order/OrderItem.java
package io.playground.orderservice.domain.order;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.math.BigDecimal;

@Getter
@AllArgsConstructor
public class OrderItem {
    private Long id;

    private Long orderId;

    private Long variantId;

    private String variantName;

    private Long productId;

    private String productName;

    private BigDecimal price;

    private int quantity;

    private OrderItemStatus status;

    private int canceledQuantity;

    private String canceledReason;

    public enum OrderItemStatus {
        NONE, PREPARING, DELIVERED, COMPLETED
    }

    public static OrderItem of(Long id,
                               Long orderId,
                               Long variantId,
                               String variantName,
                               Long productId,
                               String productName,
                               BigDecimal price,
                               int quantity,
                               OrderItemStatus status,
                               int canceledQuantity,
                               String canceledReason) {
        return new OrderItem(id, orderId, variantId, variantName, productId, productName, price, quantity, status, canceledQuantity, canceledReason);
    }
}

File: src/main/kotlin/io/playground/orderservice/domain/order/OrderItem.kt
package io.playground.orderservice.domain.order

import java.math.BigDecimal

class OrderItem(
    val id: Long?,
    val orderId: Long?,
    val variantId: Long,
    val variantName: String,
    val productId: Long,
    val productName: String,
    val price: BigDecimal,
    val quantity: Int,
    val status: OrderItemStatus,
    val canceledQuantity: Int,
    val canceledReason: String?,
) {
    enum class OrderItemStatus {
        NONE,
        PREPARING,
        DELIVERED,
        COMPLETED,
    }

    companion object {
        @JvmStatic
        fun of(
            id: Long?,
            orderId: Long?,
            variantId: Long,
            variantName: String,
            productId: Long,
            productName: String,
            price: BigDecimal,
            quantity: Int,
            status: OrderItemStatus,
            canceledQuantity: Int,
            canceledReason: String?,
        ): OrderItem = OrderItem(
            id,
            orderId,
            variantId,
            variantName,
            productId,
            productName,
            price,
            quantity,
            status,
            canceledQuantity,
            canceledReason,
        )
    }
}
